helix-view/src/decorations.rs

```diff
@@ -0,0 +1,32 @@
+use std::borrow::Cow;
+
+use crate::graphics::Style;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum TextAnnotationKind {
+    /// Add to end of line
+    Eol,
+    /// Replace actual text or arbitary cells with annotations.
+    /// Specifies an offset from the 0th column.
+    Overlay(usize),
+}
+
+impl TextAnnotationKind {
+    pub fn is_eol(&self) -> bool {
+        *self == Self::Eol
+    }
+
+    pub fn is_overlay(&self) -> bool {
+        matches!(*self, Self::Overlay(_))
+    }
+}
+
+/// Namespaces and identifes similar annotations
+pub type TextAnnotationGroup = &'static str;
```

Comment:
I think this could be an enum instead of a string, since it will be known at compile time anyway.

```rs
pub enum TextAnnotationGroup {
    Diagnostic,
    JumpMode,
}
```

Then hashing or b-tree look up & filtering will be fast, or possibly even faster would be to store annotations directly into a named field:

```rs
struct DocumentAnnotations {
    diagnostics: Vec<TextAnnotation>,
    jump_mode: Vec<TextAnnotation>,
}

pub struct Document {
    // -----8<-----
    text_annotations: DocumentAnnotations,
}
```

Reply:
ah, another review note: `Document` implements `Debug` manually, so consider adding the `text_annotations` field there as well

https://github.com/helix-editor/helix/pull/417/files#diff-36a64f6c432dc47d11c27851879a0b67459d908edfaa70262aefc21171dc62b9R129

Reply:
I think making a `TextAnnotationGroup` enum would make it hard to use annotations from a plugin, once a plugin system is implemented.

Reply:
that's a fair point! it's certainly not a blocking issue.

to reap benefits (which may be slight?) of the enum approach _and_ allow plugins to add annotations, something like this would work:

```rs
/// first-party annotation groups (only used to differentiate access to the Vecs below)
pub enum TextAnnotationGroup {
    Diagnostic,
    JumpMode,
}

/// plugin annotation groups
/// (might have to reserve "diagnostic" and "jump_mode" strs?
pub type TextAnnotationPluginGroup = &'static str;

struct DocumentAnnotations {
    diagnostics: Vec<TextAnnotation>,
    jump_mode: Vec<TextAnnotation>,
    from_plugins: BTreeMap<&'static str, TextAnnotation>,
}

pub struct Document {
    // -----8<-----
    text_annotations: DocumentAnnotations,
}
```

Reply:
opdracht